src/training/train.py

```diff
@@ -246,7 +274,8 @@ def evaluate(model, data, epoch, args, tb_writer=None, tokenizer=None):
     input_dtype = get_input_dtype(args.precision)
 
     if 'val' in data and (args.val_frequency and ((epoch % args.val_frequency) == 0 or epoch == args.epochs)):
-        assert args.geometry=='clip', "Validation loss is only implemented for plain vanilla CLIP"
+        #assert args.geometry=='clip', "Validation loss is only implemented for plain vanilla CLIP"
+        compute_and_plot_norm(all_image_features, all_text_features)
```

Comment:
`all_image_features` and `all_text_features` are undefined at this point. This should be called after that huge `with torch.no_grad():` block below. You will also need to remove the unneeded metric calculation there

(Or...if you are curious & ambitious, fix the eval loss calculation there 😅)